File: tests/java/src/T009_ConfigFileIterator.java
import gov.nasa.gsfc.gmsec.api5.Config;
import gov.nasa.gsfc.gmsec.api5.ConfigFile;
import gov.nasa.gsfc.gmsec.api5.ConfigFileIterator;
import gov.nasa.gsfc.gmsec.api5.Gmsec;
import gov.nasa.gsfc.gmsec.api5.Message;
import gov.nasa.gsfc.gmsec.api5.MessageFactory;

import gov.nasa.gsfc.gmsec.api5.util.Log;


public class T009_ConfigFileIterator extends TestCase
{
	private static ConfigFile     cfgFile    = null;
	private static MessageFactory msgFactory = null;


	public static void main(String[] args)
	{
		T009_ConfigFileIterator test = new T009_ConfigFileIterator();
		test.setDescription("Test ConfigFileIterator");
		test.run(args);
	}


    public void testCase()
		throws Exception
	{
		try
		{
			setup();

			test_iterator();
			test_reset();

			teardown();
		}
		catch (Exception e)
		{
			require(e.getMessage(), false);
		}
	}


	private void test_iterator()
		throws Exception
	{
		ConfigFileIterator iter = cfgFile.getIterator();

		Log.info("Test hasNextSubscription() and nextSubscription()");

		int numSubscriptions = 0;
		while (iter.hasNextSubscription())
		{
			ConfigFile.SubscriptionEntry entry = iter.nextSubscription();

			++numSubscriptions;

			if (numSubscriptions == 1)
			{
				check("Unexpected subscription entry 1 name", entry.getName().equals("AllMessages"));
				check("Unexpected subscription entry 1 pattern", entry.getPattern().equals("GMSEC.>"));
			}
			else
			{
				check("Unexpected subscription entry 2 name", entry.getName().equals("LogMessages"));
				check("Unexpected subscription entry 2 pattern", entry.getPattern().equals("GMSEC.*.*.MSG.LOG.+"));
			}
		}
		check("Unexpected number of subscription entries", numSubscriptions == 2);


		Log.info("Test hasNextConfig() and nextConfig()");

		int numConfigs = 0;
		while (iter.hasNextConfig())
		{
			ConfigFile.ConfigEntry entry = iter.nextConfig();

			++numConfigs;

			if (numConfigs == 1)
			{
				check("Unexpected config entry 1 name", entry.getName().equals("ActiveMQ"));
				check("Unexpected config entry mw-id", entry.getConfig().getValue("mw-id").equals("activemq39"));
				check("Unexpected config entry server", entry.getConfig().getValue("server").equals("tcp://localhost:61616"));
			}
			else
			{
				check("Unexpected config entry 2 name", entry.getName().equals("Bolt"));
				check("Unexpected config entry mw-id", entry.getConfig().getValue("mw-id").equals("bolt"));
				check("Unexpected config entry server", entry.getConfig().getValue("server").equals("localhost"));
			}
		}
		check("Unexpected number of config entries", numConfigs == 2);


		Log.info("Test hasNextMessage() and nextMessage()");

		int numMessages = 0;
		while (iter.hasNextMessage())
		{
			ConfigFile.MessageEntry entry = iter.nextMessage();

			++numMessages;

			if (numMessages == 1)
			{
				check("Unexpected message entry 1 name", entry.getName().equals("Heartbeat"));
				check("Unexpected message entry 1", entry.getMessage().toXML().equals( msgFactory.createMessage("HB").toXML() ));
			}
			else
			{
				check("Unexpected message entry 2 name", entry.getName().equals("Log"));
				check("Unexpected message entry 2", entry.getMessage().toXML().equals( msgFactory.createMessage("LOG").toXML() ));
			}
		}
		check("Unexpected number of message entries", numMessages == 2);
	}


	private void test_reset()
		throws Exception
	{
		Log.info("Test reset()");

		ConfigFileIterator iter = cfgFile.getIterator();

		check("Expected to have subscriptions", iter.hasNextSubscription());
		check("Expected to have configs", iter.hasNextConfig());
		check("Expected to have messages", iter.hasNextMessage());

		while (iter.hasNextSubscription()) iter.nextSubscription();
		while (iter.hasNextConfig()) iter.nextConfig();
		while (iter.hasNextMessage()) iter.nextMessage();

		check("Expected NOT to have additional subscriptions", iter.hasNextSubscription() == false);
		check("Expected NOT to have additional configs", iter.hasNextConfig() == false);
		check("Expected NOT to have additional messages", iter.hasNextMessage() == false);

		iter.reset();

		check("Expected to have subscriptions", iter.hasNextSubscription());
		check("Expected to have configs", iter.hasNextConfig());
		check("Expected to have messages", iter.hasNextMessage());
	}


	private static void setup()
		throws Exception
	{
		if (cfgFile == null) {
			cfgFile = new ConfigFile();
		}

		if (msgFactory == null) {
			msgFactory = MessageFactory.create();
		}

		cfgFile.addSubscriptionEntry( new ConfigFile.SubscriptionEntry("AllMessages", "GMSEC.>") );
		cfgFile.addSubscriptionEntry( new ConfigFile.SubscriptionEntry("LogMessages", "GMSEC.*.*.MSG.LOG.+") );

		cfgFile.addConfig("ActiveMQ", new Config("mw-id=activemq39 server=tcp://localhost:61616", Gmsec.DataType.KEY_VALUE_DATA));
		cfgFile.addConfig("Bolt", new Config("mw-id=bolt server=localhost", Gmsec.DataType.KEY_VALUE_DATA));

		cfgFile.addMessage("Heartbeat", msgFactory.createMessage("HB"));
		cfgFile.addMessage("Log", msgFactory.createMessage("LOG"));
	}


	private static void teardown()
	{
		MessageFactory.destroy(msgFactory);
	}
}

File: tests/csharp/src/T009_ConfigFileIterator.cs
namespace T009
{
	using GMSEC.API5;
	using GMSEC.API5.Testing.Common;
	using System;
	using System.Text;


	class T009_ConfigFileIterator : TestCase
	{
		private ConfigFile cfgFile = null;
		private Message    hbMsg   = null;
		private Message    logMsg  = null;


		static void Main(string[] args)
		{
			T009_ConfigFileIterator test = new T009_ConfigFileIterator();
			test.SetDescription("Test ConfigFileIterator");
			test.Run(args);
		}


		public override void Run()
		{
			Setup();

			Test_Iterator();
			Test_Reset();
		}


		private void Test_Iterator()
		{
			ConfigFileIterator iter = cfgFile.GetIterator();

			Log.Info("Test HasNextSubscription() and NextSubscription()");

			int numSubscriptions = 0;
			while (iter.HasNextSubscription())
			{
				SubscriptionEntry entry = iter.NextSubscription();

				++numSubscriptions;

				if (numSubscriptions == 1)
				{
					Check("Unexpected subscription entry 1 name", entry.GetName() == "AllMessages");
					Check("Unexpected subscription entry 1 pattern", entry.GetPattern() == "GMSEC.>");
				}
				else
				{
					Check("Unexpected subscription entry 2 name", entry.GetName() == "LogMessages");
					Check("Unexpected subscription entry 2 pattern", entry.GetPattern() == "GMSEC.*.*.MSG.LOG.+");
				}
			}
			Check("Unexpected number of subscription entries", numSubscriptions == 2);


			Log.Info("Test HasNextConfig() and NextConfig()");

			int numConfigs = 0;
			while (iter.HasNextConfig())
			{
				ConfigEntry entry = iter.NextConfig();

				++numConfigs;

				if (numConfigs == 1)
				{
					Check("Unexpected config entry 1 name", entry.GetName() == "ActiveMQ");
					Check("Unexpected config entry mw-id", entry.GetConfig().GetValue("mw-id") == "activemq39");
					Check("Unexpected config entry server", entry.GetConfig().GetValue("server") == "tcp://localhost:61616");
				}
				else
				{
					Check("Unexpected config entry 2 name", entry.GetName() == "Bolt");
					Check("Unexpected config entry mw-id", entry.GetConfig().GetValue("mw-id") == "bolt");
					Check("Unexpected config entry server", entry.GetConfig().GetValue("server") == "localhost");
				}
			}
			Check("Unexpected number of config entries", numConfigs == 2);


			Log.Info("Test HasNextMessage() and NextMessage()");

			int numMessages = 0;
			while (iter.HasNextMessage())
			{
				MessageEntry entry = iter.NextMessage();

				++numMessages;

				if (numMessages == 1)
				{
					Check("Unexpected message entry 1 name", entry.GetName() == "Heartbeat");
					Check("Unexpected message entry 1", entry.GetMessage().ToXML() == hbMsg.ToXML());
				}
				else
				{
					Check("Unexpected message entry 2 name", entry.GetName() == "Log");
					Check("Unexpected message entry 2", entry.GetMessage().ToXML() == logMsg.ToXML());
				}
			}
			Check("Unexpected number of message entries", numMessages == 2);
		}


		private void Test_Reset()
		{
			Log.Info("Test Reset()");

			ConfigFileIterator iter = cfgFile.GetIterator();

			Check("Expected to have subscriptions", iter.HasNextSubscription());
			Check("Expected to have configs", iter.HasNextConfig());
			Check("Expected to have messages", iter.HasNextMessage());

			while (iter.HasNextSubscription()) iter.NextSubscription();
			while (iter.HasNextConfig()) iter.NextConfig();
			while (iter.HasNextMessage()) iter.NextMessage();

			Check("Expected NOT to have additional subscriptions", iter.HasNextSubscription() == false);
			Check("Expected NOT to have additional configs", iter.HasNextConfig() == false);
			Check("Expected NOT to have additional messages", iter.HasNextMessage() == false);

			iter.Reset();

			Check("Expected to have subscriptions", iter.HasNextSubscription());
			Check("Expected to have configs", iter.HasNextConfig());
			Check("Expected to have messages", iter.HasNextMessage());
		}


		private void Setup()
		{
			cfgFile = new ConfigFile();

			cfgFile.AddSubscriptionEntry( new SubscriptionEntry("AllMessages", "GMSEC.>") );
			cfgFile.AddSubscriptionEntry( new SubscriptionEntry("LogMessages", "GMSEC.*.*.MSG.LOG.+") );

			cfgFile.AddConfig("ActiveMQ", new Config("mw-id=activemq39 server=tcp://localhost:61616", DataType.KEY_VALUE_DATA));
			cfgFile.AddConfig("Bolt", new Config("mw-id=bolt server=localhost", DataType.KEY_VALUE_DATA));

			MessageFactory msgFactory = new MessageFactory();
			SetStandardFields(msgFactory);

			hbMsg  = msgFactory.CreateMessage("HB");
			logMsg = msgFactory.CreateMessage("LOG");

			logMsg.AddField("SUBCLASS", "T009");
			logMsg.AddField("OCCURRENCE-TYPE", "TEST");
			logMsg.AddField("SEVERITY", (short) 1);

			cfgFile.AddMessage("Heartbeat", hbMsg);
			cfgFile.AddMessage("Log", logMsg);
		}
	}
}
